#Aim 
#generate lists of genes assigned to eRNA peaks via
#1 proximal gene
#2 PCHIC
# Then run XGR on the combined set of genes

library(dplyr)
library(stringr)
library(tidyr)

Okabe_Ito <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000") 


#Read in eRNA data (output from DEseq2)
eRNA.CD4 <- read.table("Inputs/eRNA.CD4.txt", header = TRUE, sep = "\t") 
eRNA.CD8 <- read.table("Inputs/eRNA.CD8.txt", header = TRUE, sep = "\t") 
eRNA.CD14 <- read.table("Inputs/eRNA.CD14.txt", header = TRUE, sep = "\t") 

#Expand first column to chr start end
eRNA.CD4 <- eRNA.CD4 %>% separate(peak.name, c("chr", "start", "end"))
eRNA.CD8 <- eRNA.CD8 %>% separate(peak.name, c("chr", "start", "end"))
eRNA.CD14 <- eRNA.CD14 %>% separate(peak.name, c("chr", "start", "end"))

#change characters to numeric
eRNA.CD4$start <- as.numeric(eRNA.CD4$start)
eRNA.CD4$end <- as.numeric(eRNA.CD4$end)
eRNA.CD8$start <- as.numeric(eRNA.CD8$start)
eRNA.CD8$end <- as.numeric(eRNA.CD8$end)
eRNA.CD14$start <- as.numeric(eRNA.CD14$start)
eRNA.CD14$end <- as.numeric(eRNA.CD14$end)


#How many peaks have FC > 1.5 and padj < 0.05?
sig_peaks <- function (df) {
    df2 <- df %>% 
        filter(log2FoldChange > 0.585 | log2FoldChange < -0.585) %>%
        filter(padj<0.05) 
    return (df2)
}

#Run function on all 3 cell type
eRNA.CD4.sig <- sig_peaks(eRNA.CD4)
eRNA.CD8.sig <- sig_peaks(eRNA.CD8)
eRNA.CD14.sig <- sig_peaks(eRNA.CD14)
dim(eRNA.CD4.sig) # 161 peaks
dim (eRNA.CD8.sig) # 59 peaks
dim (eRNA.CD14.sig) #241 peaks

#Need to use top 200 peaks for pathway analysis.

#Create a function to select top 200 peaks with FC > 1.5 and sorted by p value
Top200peaks <- function (df) {
    df2 <- df %>% 
        filter(log2FoldChange > 0.585 | log2FoldChange < -0.585) %>%
        arrange(padj) %>%
        slice_head(n = 200)
        return (df2)
}

#Run function on all 3 cell type
eRNA.CD4.Top200 <- Top200peaks(eRNA.CD4)
eRNA.CD8.Top200 <- Top200peaks(eRNA.CD8)
eRNA.CD14.Top200 <- Top200peaks(eRNA.CD14)
dim (eRNA.CD4.Top200) #200 peaks
dim (eRNA.CD8.Top200) #200 peaks
dim (eRNA.CD14.Top200) #200 peaks

#Create a bed file of Top 200 significant regions

#4th column is gene name
make_bed <- function (df) {
    df2 <- df %>% 
        dplyr::select (chr, start, end, ProxGene) %>%
        mutate_all(na_if,"") %>%
        #remove rows where col 3 == col 2
        filter (start != end)
    return (df2)
}

eRNA.CD4.Top200.bed <- make_bed(eRNA.CD4.Top200)
eRNA.CD8.Top200.bed <- make_bed(eRNA.CD8.Top200)
eRNA.CD14.Top200.bed <- make_bed(eRNA.CD14.Top200)

#4th column is ones
make_bed_2 <- function (df) {
    df2 <- df %>% 
        dplyr::select (chr, start, end) %>%
        mutate_all(na_if,"") %>%
        #remove rows where col 3 == col 2
        filter (start != end) %>%
        #add a column of ones
        mutate("1" =1)
    return (df2)
}

eRNA.CD4.Top200.bed <- make_bed_2(eRNA.CD4.Top200)
eRNA.CD8.Top200.bed <- make_bed_2(eRNA.CD8.Top200)
eRNA.CD14.Top200.bed <- make_bed_2(eRNA.CD14.Top200)

#save bed files

dir.create("bed")
write.table (eRNA.CD4.Top200.bed, "./bed/eRNA.CD4.Top200.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
write.table (eRNA.CD8.Top200.bed, "./bed/eRNA.CD8.Top200.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
write.table (eRNA.CD14.Top200.bed, "./bed/eRNA.CD14.Top200.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")

#on galahad at http://galahad.well.ox.ac.uk/UCSC/Carla_WashU/Differential_peaks/



#Get list of genes assigned by proximity
eRNA.CD4.Top200.proximity <- eRNA.CD4.Top200 %>% 
    select (ProxGene) %>% 
    rename("gene" = "ProxGene")
eRNA.CD8.Top200.proximity <- eRNA.CD8.Top200 %>% 
    select (ProxGene) %>% 
    rename("gene" = "ProxGene")
eRNA.CD14.Top200.proximity <- eRNA.CD14.Top200 %>% 
    select (ProxGene) %>% 
    rename("gene" = "ProxGene")



#Background lists of genes assigned by proximity
eRNA.CD4.background.proximity <- eRNA.CD4 %>% select (ProxGene) %>% rename("gene" = "ProxGene")
eRNA.CD8.background.proximity <- eRNA.CD8 %>% select (ProxGene) %>% rename("gene" = "ProxGene")
eRNA.CD14.background.proximity <- eRNA.CD14 %>% select (ProxGene) %>% rename("gene" = "ProxGene")

#Make a concatenated background set using rbind
eRNA.background.proximity <- rbind (eRNA.CD4.background.proximity, 
                                eRNA.CD8.background.proximity, 
                                eRNA.CD14.background.proximity)

#For ATAC & ChIP I have used the ML generated by Andy to link to genes via PCHIC
#For eRNA I want to use any eRNAs that definitely have RNA expression 
#This will therefore be a restricted ML from the DEseq2 output (I will call this eRNA_subML_CD14 etc)
#I can create this by running the make_bed_2 function on the DEseq2 output
#I will leave in the column of 1s as then I can also use it to display as a track on WashU if needed

eRNA_subML_CD4 <- make_bed_2(eRNA.CD4)
eRNA_subML_CD8 <- make_bed_2(eRNA.CD8)
eRNA_subML_CD14 <- make_bed_2(eRNA.CD14)

#Save files
write.table (eRNA_subML_CD4, "./bed/eRNA_subML_CD4.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
write.table (eRNA_subML_CD8, "./bed/eRNA_subML_CD8.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
write.table (eRNA_subML_CD14, "./bed/eRNA_subML_CD14.bed", 
             col.names = FALSE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
#Use these to perform intersection with PCHIC using bedtools on rescomp.

#Read in data of peaks linked to genes via PCHIC
eRNA.CD4.PCHIC <- read.table("Inputs/PCHIC/eRNA_subML_CD4_PCHIC.txt", header = FALSE, sep = "\t")
colnames (eRNA.CD4.PCHIC) <- c("chr", "start", "end", "1", "chr_loop", "start_loop", "end_loop", "interaction", "gene", "distance")
eRNA.CD8.PCHIC <- read.table("Inputs/PCHIC/eRNA_subML_CD8_PCHIC.txt", header = FALSE, sep = "\t")
colnames (eRNA.CD8.PCHIC) <- c("chr", "start", "end", "1", "chr_loop", "start_loop", "end_loop", "interaction", "gene", "distance")
eRNA.CD14.PCHIC <- read.table("Inputs/PCHIC/eRNA_subML_CD14_PCHIC.txt", header = FALSE, sep = "\t")
colnames (eRNA.CD14.PCHIC) <- c("chr", "start", "end", "1", "chr_loop", "start_loop", "end_loop", "interaction", "gene", "distance")

#Find Top 200 genes that have entries with PCHIC
# Make function to find Top 200 genes that have entries with PCHIC
extract.eRNA.PCHIC <- function(eRNA_Top200peaks, eRNA.ML.PCHIC) {
    eRNA_Top200_PCHIC <- inner_join(eRNA_Top200peaks, eRNA.ML.PCHIC)
    #Create a list of genes for input to XGR
    #Filter out rows where gene is "."
    eRNA_Top200_PCHIC_genes <- eRNA_Top200_PCHIC %>% filter (gene !=".")
    #split genes separated by ";" into a single list
    list <- strsplit(eRNA_Top200_PCHIC_genes$gene, ";")
    # use unlist to put each gene on a new row
    final_list <- as.data.frame(unlist(list))
    #show only unique genes
    unique_list <- unique(final_list)
    #rename columns to "gene"
    colnames(unique_list) <- c("gene")
    return(unique_list)
}

#Run function on 3 cell types
eRNA.CD4.Top200.PCHIC <- extract.eRNA.PCHIC(eRNA.CD4.Top200, eRNA.CD4.PCHIC)
eRNA.CD8.Top200.PCHIC <- extract.eRNA.PCHIC(eRNA.CD8.Top200, eRNA.CD8.PCHIC)
eRNA.CD14.Top200.PCHIC <- extract.eRNA.PCHIC(eRNA.CD14.Top200, eRNA.CD14.PCHIC)
dim(eRNA.CD4.Top200.PCHIC) #214 genes
dim(eRNA.CD8.Top200.PCHIC) #185 genes
dim(eRNA.CD14.Top200.PCHIC) #176 genes

#Generate background set for PCHIC linked genes
extract.eRNA.background.PCHIC <- function (eRNA.ML.PCHIC){
    #Filter out rows where gene is "."
    eRNA.ML.PCHIC.genes <- eRNA.ML.PCHIC %>% filter (gene !=".")
    #split genes separated by ";" into a single list
    list <- strsplit(eRNA.ML.PCHIC.genes$gene, ";")
    # use unlist to put each gene on a new row
    final_list <- as.data.frame(unlist(list))
    #show only unique genes
    unique_list <- unique(final_list)
    #rename columns to "gene"
    colnames(unique_list) <- c("gene")
    return(unique_list)
}

eRNA.CD4.background.PCHIC <- extract.eRNA.background.PCHIC(eRNA.CD4.PCHIC)
eRNA.CD8.background.PCHIC <- extract.eRNA.background.PCHIC(eRNA.CD8.PCHIC)
eRNA.CD14.background.PCHIC <- extract.eRNA.background.PCHIC(eRNA.CD14.PCHIC)
#Make a concatenated background set using rbind
eRNA.background.PCHIC <- rbind (eRNA.CD4.background.PCHIC, 
                                eRNA.CD8.background.PCHIC, 
                                eRNA.CD14.background.PCHIC)

#Generate concatenated lists of genes from both methods
eRNA.CD4.genes <- rbind (eRNA.CD4.Top200.proximity, eRNA.CD4.Top200.PCHIC)
eRNA.CD8.genes <- rbind (eRNA.CD8.Top200.proximity, eRNA.CD8.Top200.PCHIC)
eRNA.CD14.genes <- rbind (eRNA.CD14.Top200.proximity, eRNA.CD14.Top200.PCHIC)
#Convert these lists to vectors
eRNA.CD4.genes <- as.vector(eRNA.CD4.genes$gene)
eRNA.CD8.genes <- as.vector(eRNA.CD8.genes$gene)
eRNA.CD14.genes <- as.vector(eRNA.CD14.genes$gene)
length(eRNA.CD4.genes) #414 genes
length(eRNA.CD8.genes) #385 genes
length(eRNA.CD14.genes) #376
#So I can use the default parameters for XGR for all 3 cell types

#Generate concatenated background for all 3 cell types
eRNA.background <- rbind(eRNA.background.proximity, eRNA.background.PCHIC)
eRNA.background <- as.vector(eRNA.background$gene)
class(eRNA.background)

library(XGR)
dir.create("GOBP")
dir.create("Reactome")
RData.location <- "http://galahad.well.ox.ac.uk/bigdata"

XGR_function <- function(genelist, background, ontology) {
    #Run xEnricher function 
    #NB settings: for <50 genes use size.range = c(5,2000) min.overlap = 3
    #For larger gene sets than 50 use size.range = c(10,2000) min.overlap = 5 (which is the default setting)
    enriched_genes <- xEnricherGenes(data = genelist, 
                                     background = background, 
                                     test = "hypergeo", 
                                     ontology = ontology, 
                                     RData.location = RData.location)
    enriched_genes_concise <- xEnrichConciser(enriched_genes)
    return (enriched_genes_concise)
}

eRNA.CD4.GOBP <- XGR_function(eRNA.CD4.genes, eRNA.background, ontology = "GOBP")
eRNA.CD4.GOBP.table <- xEnrichViewer(eRNA.CD4.GOBP, details = TRUE, 
                                    top_num = 50)
write.table(eRNA.CD4.GOBP.table, "./GOBP/eRNA.CD4.GOBP.txt", sep = "\t", quote = FALSE)

eRNA.CD8.GOBP <- XGR_function(eRNA.CD8.genes, eRNA.background, ontology = "GOBP")
eRNA.CD8.GOBP.table <- xEnrichViewer(eRNA.CD8.GOBP, details = TRUE, 
                                     top_num = 50)
write.table(eRNA.CD8.GOBP.table, "./GOBP/eRNA.CD8.GOBP.txt", sep = "\t", quote = FALSE)

eRNA.CD14.GOBP <- XGR_function(eRNA.CD14.genes, eRNA.background, ontology = "GOBP")
eRNA.CD14.GOBP.table <- xEnrichViewer(eRNA.CD14.GOBP, details = TRUE, 
                                     top_num = 50)
write.table(eRNA.CD14.GOBP.table, "./GOBP/eRNA.CD14.GOBP.txt", sep = "\t", quote = FALSE)

#make a list of things to compare
eRNA.GOBP <- list(eRNA.CD4.GOBP, eRNA.CD8.GOBP, eRNA.CD14.GOBP)
names(eRNA.GOBP) <- c("CD4 T cells", "CD8 T cells", "CD14 monocytes")

#Set the order
eRNA.GOBP <- eRNA.GOBP[c("CD14 monocytes", "CD8 T cells", "CD4 T cells")]

eRNA.GOBP.compare <- xEnrichCompare(eRNA.GOBP, FDR.cutoff = 0.05, facet = TRUE, signature = FALSE,
                                    bar.label.size = 4)
eRNA.GOBP.compare <- 
    eRNA.GOBP.compare + theme(axis.text.y=element_text(size=20,color="black")) +
    scale_fill_manual(values = c(Okabe_Ito[3], Okabe_Ito[5], Okabe_Ito[6])) 
eRNA.GOBP.compare

dir.create("figures", showWarnings = FALSE)
ggsave("eRNA.GOBP.pdf", eRNA.GOBP.compare, width = 15, height = 10, useDingbats = FALSE, path = "./figures/")
ggsave("eRNA.GOBP.png", eRNA.GOBP.compare, width = 15, height = 10, device = "png", path = "./figures/")




#Now do for Reactome

eRNA.CD4.Reactome <- XGR_function(eRNA.CD4.genes, eRNA.background, ontology = "MsigdbC2REACTOME")

eRNA.CD4.Reactome.table <- xEnrichViewer(eRNA.CD4.Reactome, details = TRUE, 
                                     top_num = 50)

write.table(eRNA.CD4.Reactome.table, "./Reactome/eRNA.CD4.Reactome.txt", sep = "\t", quote = FALSE)

eRNA.CD8.Reactome <- XGR_function(eRNA.CD8.genes, eRNA.background, ontology = "MsigdbC2REACTOME")
eRNA.CD8.Reactome.table <- xEnrichViewer(eRNA.CD8.Reactome, details = TRUE, 
                                     top_num = 50)
write.table(eRNA.CD8.Reactome.table, "./Reactome/eRNA.CD8.Reactome.txt", sep = "\t", quote = FALSE)

eRNA.CD14.Reactome <- XGR_function(eRNA.CD14.genes, eRNA.background, ontology = "MsigdbC2REACTOME")
eRNA.CD14.Reactome.table <- xEnrichViewer(eRNA.CD14.Reactome, details = TRUE, 
                                      top_num = 50)
write.table(eRNA.CD14.Reactome.table, "./Reactome/eRNA.CD14.Reactome.txt", sep = "\t", quote = FALSE)

#Perform comparison
eRNA.Reactome <- list(eRNA.CD4.Reactome, eRNA.CD8.Reactome, eRNA.CD14.Reactome)
names(eRNA.Reactome) <- c("CD4+ T cells", "CD8+ T cells", "CD14+ monocytes")
#Set the order
eRNA.Reactome <- eRNA.Reactome[c("CD14+ monocytes", "CD8+ T cells", "CD4+ T cells")]

#Perform the comparison
eRNA.Reactome.compare <- xEnrichCompare(eRNA.Reactome, FDR.cutoff = 0.05, 
                                           signature = FALSE,
                                           bar.label.size = 4)

eRNA.Reactome.compare <- 
    eRNA.Reactome.compare + theme(axis.text.y=element_text(size=24,color="black")) +
    scale_fill_manual(values = c(Okabe_Ito[3], Okabe_Ito[5], Okabe_Ito[6])) 
eRNA.Reactome.compare

dir.create("figures", showWarnings = FALSE)
ggsave("eRNA.Reactome.pdf", eRNA.Reactome.compare, width = 15, height = 10, useDingbats = FALSE, path = "./figures/")
ggsave("eRNA.Reactome.png", eRNA.Reactome.compare, width = 15, height = 10, device = "png", path = "./figures/")


#Repeat XGR analysis with less stringent parameters for Reactome
XGR_function_2 <- function(genelist, background, ontology) {
    #Run xEnricher function 
    #NB settings: for <50 genes use size.range = c(5,2000) min.overlap = 3
    #For larger gene sets than 50 use size.range = c(10,2000) min.overlap = 5 (which is the default setting)
    enriched_genes <- xEnricherGenes(data = genelist, 
                                     background = background, 
                                     test = "hypergeo", 
                                     ontology = ontology, 
                                     RData.location = RData.location, 
                                     size.range = c(5,2000), 
                                     min.overlap = 3)
    enriched_genes_concise <- xEnrichConciser(enriched_genes)
    return (enriched_genes_concise)
}

eRNA.CD4.Reactome_2 <- XGR_function_2(eRNA.CD4.genes, eRNA.background, ontology = "MsigdbC2REACTOME")

eRNA.CD4.Reactome.table_2 <- xEnrichViewer(eRNA.CD4.Reactome_2, details = TRUE, 
                                         top_num = 50)

write.table(eRNA.CD4.Reactome.table_2, "./Reactome/eRNA.CD4.Reactome_2.txt", sep = "\t", quote = FALSE)

eRNA.CD8.Reactome_2 <- XGR_function_2(eRNA.CD8.genes, eRNA.background, ontology = "MsigdbC2REACTOME")

eRNA.CD8.Reactome.table_2 <- xEnrichViewer(eRNA.CD8.Reactome_2, details = TRUE, 
                                           top_num = 50)

write.table(eRNA.CD8.Reactome.table_2, "./Reactome/eRNA.CD4.Reactome_2.txt", sep = "\t", quote = FALSE)

eRNA.CD14.Reactome_2 <- XGR_function_2(eRNA.CD14.genes, eRNA.background, ontology = "MsigdbC2REACTOME")

eRNA.CD14.Reactome.table_2 <- xEnrichViewer(eRNA.CD14.Reactome_2, details = TRUE, 
                                           top_num = 50)

write.table(eRNA.CD14.Reactome.table_2, "./Reactome/eRNA.CD14.Reactome_2.txt", sep = "\t", quote = FALSE)


#Perform comparison
eRNA.Reactome_2 <- list(eRNA.CD4.Reactome_2, eRNA.CD8.Reactome_2, eRNA.CD14.Reactome_2)
names(eRNA.Reactome_2) <- c("CD4+ T cells", "CD8+ T cells", "CD14+ monocytes")
#Set the order
eRNA.Reactome_2 <- eRNA.Reactome_2[c("CD14+ monocytes", "CD8+ T cells", "CD4+ T cells")]

#Perform the comparison
eRNA.Reactome.compare_2 <- xEnrichCompare(eRNA.Reactome_2, FDR.cutoff = 0.05, 
                                        signature = FALSE,
                                        bar.label.size = 4)

eRNA.Reactome.compare_2 <- 
    eRNA.Reactome.compare_2 + theme(axis.text.y=element_text(size = 16, color="black")) +
    scale_fill_manual(values = c(Okabe_Ito[3], Okabe_Ito[5], Okabe_Ito[6])) 
eRNA.Reactome.compare_2

dir.create("figures", showWarnings = FALSE)
ggsave("eRNA.Reactome_2.pdf", eRNA.Reactome.compare_2, width = 15, height = 10, useDingbats = FALSE, path = "./figures/")
ggsave("eRNA.Reactome_2.png", eRNA.Reactome.compare_2, width = 15, height = 10, device = "png", path = "./figures/")


#Find examples from output 
#e.g. from "Signalling by NOTCH" category in CD14
#	B4GALT1, MFNG, NCOR2, RBX1, TBL1XR1, TLE3


#B4GALT1
#Is B4GALT1 in the list of PCHIC genes or proximity genes?
eRNA.CD14.Top200.PCHIC %>% filter (gene == "B4GALT1")
#Yes
eRNA.CD14.Top200.proximity %>% filter (gene == "B4GALT1")
#No

#What is the p value of that DE region?
#Join PCHIC and DEseq2 results
eRNA.CD14.PCHIC %>% inner_join(eRNA.CD14) %>%
    #filter gene of interest
    filter (grepl('B4GALT1', gene))
#p = 0.00122 looks like a good one to use

#MFNG
#Is MFNG in the list of PCHIC genes or proximity genes?
eRNA.CD14.Top200.PCHIC %>% filter (gene == "MFNG")
#No
eRNA.CD14.Top200.proximity %>% filter (gene == "MFNG")
#Yes

#What is the p value of that DE region?
eRNA.CD14 %>%
    #filter gene of interest
    filter (grepl('MFNG', ProxGene))
#p = 0.0002 looks like a good one to use

#NCOR2
#Is NCOR2 in the list of PCHIC genes or proximity genes?
eRNA.CD14.Top200.PCHIC %>% filter (gene == "NCOR2")
#Yes
eRNA.CD14.Top200.proximity %>% filter (gene == "NCOR2")
#No

#What is the p value of that DE region?
#Join PCHIC and DEseq2 results
eRNA.CD14.PCHIC %>% inner_join(eRNA.CD14) %>%
    #filter gene of interest
    filter (grepl('NCOR2', gene)) %>%
    arrange(padj)
    
#chr12 125197111 125199111
#Good example with three DE eRNA in long enhancer



#Need to do some visualisations one tracks in place

###Are any regions near GWAS loci?### Another method
#Used bedtools to look for any eRNAs within 500kb of a lead SNP
#Results are in ./peak_vs_SNP

#Read in the data

eRNA_ML_CD4.SNP <- read.table("./peak_vs_SNP/eRNA_subML_CD4.SNP.txt", header = FALSE, sep = "\t")
colnames (eRNA_ML_CD4.SNP) <- c("chr.SNP", "start.SNP", "end.SNP", "SNP", 
                                   "chr", "start", "end", "1")
eRNA_ML_CD8.SNP <- read.table("./peak_vs_SNP/eRNA_subML_CD8.SNP.txt", header = FALSE, sep = "\t")
colnames (eRNA_ML_CD8.SNP) <- c("chr.SNP", "start.SNP", "end.SNP", "SNP", 
                                   "chr", "start", "end", "1")
eRNA_ML_CD14.SNP <- read.table("./peak_vs_SNP/eRNA_subML_CD14.SNP.txt", header = FALSE, sep = "\t")
colnames (eRNA_ML_CD14.SNP) <- c("chr.SNP", "start.SNP", "end.SNP", "SNP", 
                                    "chr", "start", "end", "1")

#Join with DEseq2 output, select peaks with FC > 1.5, p<0.05 and sort by distance to SNP

SNPs_peaks <- function (df, DEseq2) {
    df2 <- df %>% left_join(DEseq2) %>%
        filter (padj < 0.05) %>%
        filter(log2FoldChange > 0.585 | log2FoldChange < -0.585) %>%
        mutate (snp_dist = end.SNP - end) 
    #change snp dist to always be positive
    df2$snp.dist <- abs(df2$snp_dist)    
    df2 <-  df2 %>% arrange (snp.dist) %>%
        select (SNP, chr, start, end, baseMean, log2FoldChange, pvalue, padj, ProxGene, snp.dist)
    return (df2)
}

eRNA.CD4.SNP <- SNPs_peaks(eRNA_ML_CD4.SNP, eRNA.CD4)
eRNA.CD8.SNP <- SNPs_peaks(eRNA_ML_CD8.SNP, eRNA.CD8)
eRNA.CD14.SNP <- SNPs_peaks(eRNA_ML_CD14.SNP, eRNA.CD14)

#Lots of interesting examples to check out

write.table (eRNA.CD4.SNP, "./peak_vs_SNP/eRNA.CD4.SNP.txt", 
             col.names = TRUE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")

write.table (eRNA.CD8.SNP, "./peak_vs_SNP/eRNA.CD8.SNP.txt", 
             col.names = TRUE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")
write.table (eRNA.CD14.SNP, "./peak_vs_SNP/eRNA.CD14.SNP.txt", 
             col.names = TRUE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")

#Are any eRNA differential peaks at DE genes?
#Read in gene expression data

#Import DEseq2 data from RNA-seq

RNA.CD4 <- read.table("C:/Users/ccohen.WHGC/Documents/R/XGR/RNAseq_cohort2-5/Inputs/RNA.CD4.txt",
                      header = TRUE, sep = "\t") 
RNA.CD8 <- read.table("C:/Users/ccohen.WHGC/Documents/R/XGR/RNAseq_cohort2-5/Inputs/RNA.CD8.txt",
                      header = TRUE, sep = "\t") 
RNA.CD14 <- read.table("C:/Users/ccohen.WHGC/Documents/R/XGR/RNAseq_cohort2-5/Inputs/RNA.CD14.txt",
                       header = TRUE, sep = "\t") 
#Intersect Top 200 eRNA peaks with RNAseq data and select RNA >1.5 fold change and padj<0.05
CD4.eRNA.RNA.sig <- eRNA.CD4.Top200 %>% inner_join(RNA.CD4, by = c("ProxGene" = "name")) %>%
    filter(log2FoldChange.y > 0.585 | log2FoldChange.y < -0.585) %>%
    filter(padj.y<0.05) 
CD8.eRNA.RNA.sig <- eRNA.CD8.Top200 %>% inner_join(RNA.CD8, by = c("ProxGene" = "name")) %>%
    filter(log2FoldChange.y > 0.585 | log2FoldChange.y < -0.585) %>%
    filter(padj.y<0.05) 
CD14.eRNA.RNA.sig <- eRNA.CD14.Top200 %>% inner_join(RNA.CD14, by = c("ProxGene" = "name")) %>%
    filter(log2FoldChange.y > 0.585 | log2FoldChange.y < -0.585) %>%
    filter(padj.y<0.05) 
dim(CD14.eRNA.RNA.sig) # 8 regions
dim (CD4.eRNA.RNA.sig)#4 regions
dim (CD8.eRNA.RNA.sig) #6 regions
dir.create("Examples")
write.table (CD14.eRNA.RNA.sig, "./Examples/CD14.eRNA.RNA.sig.txt", 
             col.names = TRUE,
             row.names = FALSE,
             quote = FALSE,
             na = "NA",
             sep = "\t")



